Enable preemption and borrowing based on flavor

KunWuLuan opened the pull request:
/kind feature

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #582 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Options to allow workloads to borrow quota or preempt other workloads before trying the next flavor in the list
```

alculquicondor commented:
/ok-to-test

alculquicondor commented:
/assign @trasc for a first pass

trasc commented:
/assign

trasc commented on pkg/cache/cache.go:
```diff
@@ -140,6 +141,7 @@ type ClusterQueue struct {
 	NamespaceSelector labels.Selector
 	Preemption        kueue.ClusterQueuePreemption
 	Status            metrics.ClusterQueueStatus
+	FlavorFungibility v1beta1.FlavorFungibility
```
Where is this populated?

KunWuLuan replied:
I removed this field, we can get this from cq.Preemption

trasc commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -389,18 +392,39 @@ func (a *Assignment) findFlavorForResourceGroup(
-				// All the resources fit in the cohort, no need to check more flavors.
-				return bestAssignment, nil
-			}
+			// if bestAssignmentMode == Fit {
+			// 	// All the resources fit in the cohort, no need to check more flavors.
+			// 	return bestAssignment, nil
+			// }
```
```suggestion
```

KunWuLuan replied:
Done, thanks

trasc reviewed (commented):
Please try to fix the tests.

alculquicondor commented on apis/kueue/v1beta1/clusterqueue_types.go:
```diff
@@ -243,6 +243,21 @@ const (
+)
+
+type FlavorFungibility struct {
+	// +kubebuilder:validation:Enum="Borrow,TryNextFlavor"
```
set defaults

KunWuLuan replied:
done

alculquicondor commented on apis/kueue/v1beta1/clusterqueue_types.go:
```diff
@@ -276,6 +291,8 @@ type ClusterQueuePreemption struct {
 	// +kubebuilder:validation:Enum=Never;LowerPriority;LowerOrNewerEqualPriority
 	WithinClusterQueue PreemptionPolicy `json:"withinClusterQueue,omitempty"`
+
+	FlavorFungibility FlavorFungibility `json:"flavorFungibility"`
```
beyond prototype: needs an explanation comment.

alculquicondor commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -389,18 +392,39 @@ func (a *Assignment) findFlavorForResourceGroup(
 		if representativeMode > bestAssignmentMode {
 			bestAssignment = assignments
 			bestAssignmentMode = representativeMode
-			if bestAssignmentMode == Fit {
```
why is it ok to remove this?

KunWuLuan replied:
we will handle this case in shouldTryNextFlavor like:
```
	if representativeMode == Fit && whetherNeedBorrowing && policyBorrow == v1beta1.Borrow {
		return false
	}

	if representativeMode == Fit && !whetherNeedBorrowing {
		return false
	}
```

KunWuLuan replied:
When representativeMode == Fit and whetherNeedBorrowing == true and policyBorrow == v1beta1.Borrow, we will visit next flavor so bestAssignmentMode == Fit should not be break the loop

alculquicondor commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -389,18 +392,39 @@ func (a *Assignment) findFlavorForResourceGroup(
+	policyPreempt := flavorFungibility.WhenCanPreempt
+	policyBorrow := flavorFungibility.WhenCanBorrow
+	if representativeMode == Preempt && policyPreempt == v1beta1.Preempt {
+		return true
```
false?

KunWuLuan replied:
fixed

alculquicondor commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -389,18 +392,39 @@ func (a *Assignment) findFlavorForResourceGroup(
+func shouldTryNextFlavor(representativeMode FlavorAssignmentMode, flavorFungibility v1beta1.FlavorFungibility, whetherNeedBorrowing bool) bool {
+	policyPreempt := flavorFungibility.WhenCanPreempt
+	policyBorrow := flavorFungibility.WhenCanBorrow
+	if representativeMode == Preempt && policyPreempt == v1beta1.Preempt {
```
at this point, we don't know if preemption would be possible (there might not be enough candidates for preemption)

So how would we "try the next flavor" if preemption doesn't help.

Another thought that comes to mind is whether we should prioritize partial admission over "trying the next flavor".

I think we need more thinking about policies overall.

alculquicondor replied:
Another possible implementation path is that we somehow keep some information in memory about the last flavors and modes we used. This way, in a future scheduling attempt, we can try the next flavor if preemption didn't work.

KunWuLuan replied:
Yes, it is better to try next flavor the next time workload been scheduled. I will try to give some implementation for this.

trasc replied:
Has this changed?

KunWuLuan commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -389,18 +392,39 @@ func (a *Assignment) findFlavorForResourceGroup(
 			}
 		}
 
+		if shouldTryNextFlavor(representativeMode, cq.FlavorFungibility, whetherNeedBorrowing) {
```
yes, thanks

KunWuLuan commented:
/test all

KunWuLuan commented:
/retest

KunWuLuan commented:
/retest

KunWuLuan commented:
/retest

KunWuLuan commented:
@alculquicondor 🙇 Spent a lot of time fixing tests that couldn't pass. If you have time, please take a look, thanks!

trasc reviewed (commented):
Is the new functionality tested?

KunWuLuan commented on pkg/scheduler/scheduler_test.go:
```diff
@@ -197,97 +196,97 @@ func TestSchedule(t *testing.T) {
-				"eng-alpha": sets.New("sales/new"),
-			},
-		},
+		// "workload fits in single clusterQueue": {
```
I forgot to uncomment these, sorry. I have fixed this.

KunWuLuan commented:
> Is the new functionality tested?

Not yet, I will add some tests this week

KunWuLuan commented:
@trasc Hi, I have add some tests for this new func, if you have time, please take a look. Thanks very much!

trasc reviewed (commented):
The tests are fine, but I  don't think that the case in which the preemption cannot be done is covered.

KunWuLuan commented:
> The tests are fine, but I don't think that the case in which the preemption cannot be done is covered.

OK, will work on adding test to cover this case

KunWuLuan commented:
> The tests are fine, but I don't think that the case in which the preemption cannot be done is covered.

@trasc Hi, I have add a unit test in scheduler's tests for this case. In this test, if we don't start where preempting failed, workload `default/preemptor` will always try to preempt in the first flavor. So if the workload was admitted and current workload is not been preempted, we can say the scheduler's behavior is not wrong

trasc commented on pkg/scheduler/scheduler_test.go:
```diff
@@ -1070,6 +1070,177 @@ func TestEntryOrdering(t *testing.T) {
 	}
 }
 
+func TestLastSchedulingContext(t *testing.T) {
```
Can you add more test scenarios?

With more than  two flavors.
When the admission is not possible (get to the end of the flavor list) 
What happens when the queue changes configuration (adds / removes flavors) between admission attempts.

KunWuLuan replied:
@trasc Hi, I have add some more test cases, including some test cases for scheduling context changed and for borrow before next flavor and borrow after all flavors. Please take a look, thank you!

KunWuLuan commented:
@trasc Hi, if you have time, please take a look at this pr again, thanks. :D

trasc commented:
> @trasc Hi, if you have time, please take a look at this pr again, thanks. :D

@KunWuLuan can you mark the resolved points accordingly and maybe retitle the PR if it is no longer WIP?

KunWuLuan commented on apis/kueue/v1beta1/clusterqueue_types.go:
```diff
@@ -243,6 +243,21 @@ const (
+	// +kubebuilder:validation:Enum="Preempt,TryNextFlavor"
+	WhenCanPreempt FlavorFungibilityPolicy `json:"whenCanPreempt"`
+}
+
```
done

KunWuLuan commented:
> > @trasc Hi, if you have time, please take a look at this pr again, thanks. :D
> 
> @KunWuLuan can you mark the resolved points accordingly and maybe retitle the PR if it is no longer WIP?

@trasc  Done

alculquicondor commented:
I've been meaning to get back to this but I haven't had the chance.

Could you also update the related KEP to match the implementation so I can more easily get a high level idea of how you are trying to achieve this feature?

Also FYI on #994, which might greatly affect this

KunWuLuan commented:
> I've been meaning to get back to this but I haven't had the chance.
> 
> Could you also update the related KEP to match the implementation so I can more easily get a high level idea of how you are trying to achieve this feature?
> 
> Also FYI on #994, which might greatly affect this

@alculquicondor Done. I think this pr is nothing to do with https://github.com/kubernetes-sigs/kueue/pull/994, cause we will only decide whether a flavor should be considered and will not influence the preemption and admission any more.

alculquicondor commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -218,33 +220,67 @@ func (m FlavorAssignmentMode) String() string {
+	}
+	for flavor, flavorusage := range cq.Usage {
+		assignment.ScheduleState.ClusterQueueUsage[flavor] = make(map[corev1.ResourceName]int64)
+		for res, usage := range flavorusage {
+			assignment.ScheduleState.ClusterQueueUsage[flavor][res] = usage
+		}
```
can you use a single maps.Clone?

alculquicondor commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -218,33 +220,67 @@ func (m FlavorAssignmentMode) String() string {
+		assignment.ScheduleState.ResourceFlavors[res] = flavor
+	}
+	for flavor, flavorusage := range cq.Usage {
+		assignment.ScheduleState.ClusterQueueUsage[flavor] = make(map[corev1.ResourceName]int64)
```
```suggestion
		assignment.ScheduleState.ClusterQueueUsage[flavor] = make(map[corev1.ResourceName]int64, len(flavorusage))
```

alculquicondor commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -218,33 +220,67 @@ func (m FlavorAssignmentMode) String() string {
+	}
+	if cq.Cohort != nil {
+		for flavor, flavorusage := range cq.Cohort.Usage {
+			assignment.ScheduleState.CohortUsage[flavor] = make(map[corev1.ResourceName]int64)
```
Same recommendations as above.

alculquicondor commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -218,33 +220,67 @@ func (m FlavorAssignmentMode) String() string {
 		TotalBorrow: make(cache.FlavorResourceQuantities),
 		PodSets:     make([]PodSetAssignment, 0, len(requests)),
-		usage:       make(cache.FlavorResourceQuantities),
+		ScheduleState: workload.LastScheduleClusterQueueState{
```
if there is a previous state, we already verified that it matches the previous state of the CQ and cohort. So in that case, we shouldn't need to create new maps, we can reuse the existing ones.

KunWuLuan replied:
done

alculquicondor commented on pkg/workload/workload.go:
```diff
@@ -39,6 +40,40 @@ var (
+}
+
 // Info holds a Workload object and some pre-processing.
 type Info struct {
```
Not sure if this will be relevant, but worth checking.

When a admission attempt fails, we usually send a status update to the workload with an Admitted=False condition

https://github.com/kubernetes-sigs/kueue/blob/3d1dc9a4fb78c52bdb8ca53d8c94139a6f241304/pkg/scheduler/scheduler.go#L470-L471

The update to the object would cause an update into the queue:

https://github.com/kubernetes-sigs/kueue/blob/3d1dc9a4fb78c52bdb8ca53d8c94139a6f241304/pkg/controller/core/workload_controller.go#L297

The problem is that in there we always call `workload.NewInfo`, which could wipe the saved status.

Note that requeuing happens before sending the update https://github.com/kubernetes-sigs/kueue/blob/3d1dc9a4fb78c52bdb8ca53d8c94139a6f241304/pkg/scheduler/scheduler.go#L466-L471

But with enough retries, it is possible that a Workload update comes while we are doing a scheduling attempt. Then, we might need to change `requeueIfNotPresent` to still update the lastState.

https://github.com/kubernetes-sigs/kueue/blob/3d1dc9a4fb78c52bdb8ca53d8c94139a6f241304/pkg/queue/cluster_queue_impl.go#L130

KunWuLuan replied:
I will add some check in `Manager.addOrUpdateWorkload()` to judge whether to copy lastState

alculquicondor commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -271,7 +307,11 @@ func assignFlavors(log logr.Logger, requests []workload.PodSetResources, podSets
 				break
 			}
-			flavors, status := assignment.findFlavorForResourceGroup(log, rg, podSet.Requests, resourceFlavors, cq, &podSets[i].Template.Spec)
+			var lastScheduleFlavorInfo map[corev1.ResourceName]int
```
```suggestion
			var lastFlavorAssignment map[corev1.ResourceName]int
```

alculquicondor commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -271,7 +307,11 @@ func assignFlavors(log logr.Logger, requests []workload.PodSetResources, podSets
-			flavors, status := assignment.findFlavorForResourceGroup(log, rg, podSet.Requests, resourceFlavors, cq, &podSets[i].Template.Spec)
+			var lastScheduleFlavorInfo map[corev1.ResourceName]int
+			if lastSchedule != nil {
+				lastScheduleFlavorInfo = lastSchedule.LastScheduledFlavorIdx[podSet.Name]
```
```suggestion
				lastScheduleFlavorInfo = lastSchedule.LastScheduledFlavorIdx[i]
```

Better to use slices, as they are faster to allocate and query.

KunWuLuan replied:
done, thanks

alculquicondor commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -305,6 +345,7 @@ func (psa *PodSetAssignment) append(flavors ResourceAssignment, status *Status)
 }
 
 func (a *Assignment) append(requests workload.Requests, psAssignment *PodSetAssignment) {
+	var flavorIdx = map[corev1.ResourceName]int{}
```
```suggestion
	var flavorIdx = make(map[corev1.ResourceName]int, len(psAssignment.Flavors))
```

alculquicondor commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -364,10 +410,17 @@ func (a *Assignment) findFlavorForResourceGroup(
 			continue
 		}
 
+		whetherNeedBorrowing := false
```
```suggestion
		needsBorrowing := false
```

alculquicondor commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -341,7 +385,9 @@ func (a *Assignment) findFlavorForResourceGroup(
 	// We will only check against the flavors' labels for the resource.
 	selector := flavorSelector(spec, rg.LabelKeys)
-	for _, flvQuotas := range rg.Flavors {
+	lastScheduledFlavorIdx := 0
```
Within a resource group, all the resources get assigned the same flavor.
So at this point you can find the last flavor used for any of the resources in this resource group, and start the iteration from there.

It would be the same behavior you have below, but I think it will be easier to read.

alculquicondor commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -388,19 +442,52 @@ func (a *Assignment) findFlavorForResourceGroup(
+		return false
+	}
+
+	if representativeMode == Fit && whetherNeedBorrowing && policyBorrow == v1beta1.Borrow {
```
Is the code any simpler if we make Borrow be a mode?

alculquicondor reviewed (commented):
I'm overall happy with the approach.

Still, it might be worth adding a feature gate to disable the feature in case we find problems. Could you do that?

trasc commented on pkg/workload/workload.go:
```diff
@@ -39,6 +40,40 @@ var (
+	CohortUsage            FlavorResourceQuantities
+}
+
+func (s *LastScheduleClusterQueueState) Equal(resourceFlavors map[kueue.ResourceFlavorReference]*kueue.ResourceFlavor, clusterQueueUsage, cohortUsage FlavorResourceQuantities) bool {
```
If going with this full snapshot approach, can't we do more with that data, so the assignment from the beginning is triggered only the current state of the queue/cohort gives better chances to fit in a "better" flavor.

KunWuLuan replied:
I will record the generation of the Cohort and cluster queue, and increase the generation when more resources were add to cohort or cluster queue (either some workloads were released or cluster queue/cohort was changed)

KunWuLuan commented on apis/kueue/v1beta1/clusterqueue_types.go:
```diff
@@ -276,6 +295,17 @@ type ClusterQueuePreemption struct {
 	// +kubebuilder:validation:Enum=Never;LowerPriority;LowerOrNewerEqualPriority
 	WithinClusterQueue PreemptionPolicy `json:"withinClusterQueue,omitempty"`
+
+	// flagorFungibility determines whether a workload should try the next flavor
```
done

KunWuLuan commented on pkg/workload/workload.go:
```diff
@@ -39,6 +40,40 @@ var (
 )
 
+type FlavorResourceQuantities map[kueue.ResourceFlavorReference]map[corev1.ResourceName]int64
+type LastScheduleClusterQueueState struct {
```
done

KunWuLuan commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -218,33 +220,67 @@ func (m FlavorAssignmentMode) String() string {
 // FlavorAssignmentMode.
 func AssignFlavors(log logr.Logger, wl *workload.Info, resourceFlavors map[kueue.ResourceFlavorReference]*kueue.ResourceFlavor, cq *cache.ClusterQueue, counts []int32) Assignment {
+	if wl.LastSchedule != nil {
+		if cq.Cohort == nil && !wl.LastSchedule.Equal(resourceFlavors, workload.FlavorResourceQuantities(cq.Usage), make(workload.FlavorResourceQuantities)) {
```
done

KunWuLuan commented on pkg/scheduler/scheduler.go:
```diff
@@ -263,6 +263,7 @@ func (s *Scheduler) nominate(ctx context.Context, workloads []workload.Info, sna
 		} else {
 			e.assignment, e.preemptionTargets = s.getAssignments(log, &e.Info, &snap)
 			e.inadmissibleMsg = e.assignment.Message()
+			e.Info.LastSchedule = e.assignment.ScheduleState.Clone()
```
done,thanks

KunWuLuan commented on pkg/workload/workload.go:
```diff
@@ -39,6 +40,40 @@ var (
 	admissionManagedConditions = []string{kueue.WorkloadAdmitted, kueue.WorkloadEvicted}
 )
 
+type FlavorResourceQuantities map[kueue.ResourceFlavorReference]map[corev1.ResourceName]int64
```
done, there will an import cycle if using cache.FlavorResourceQuantities.
I have move definition from package cache to package workload

alculquicondor replied:
Did you though about the possibility of reverting this?

tenzen-y replied:
@KunWuLuan ping

KunWuLuan replied:
done. thanks for reminding me

KunWuLuan commented:
> I'm overall happy with the approach.
> 
> Still, it might be worth adding a feature gate to disable the feature in case we find problems. Could you do that?

Have added a feature gate

KunWuLuan commented on pkg/workload/workload.go:
```diff
@@ -39,6 +40,40 @@ var (
+	LastScheduledFlavorIdx map[string]map[corev1.ResourceName]int
+	ResourceFlavors        map[kueue.ResourceFlavorReference]*kueue.ResourceFlavor
+	ClusterQueueUsage      FlavorResourceQuantities
+	CohortUsage            FlavorResourceQuantities
```
done

KunWuLuan commented:
@alculquicondor @trasc Hi, I think this pr is ready for review again. Thanks very much if u have time to take a look.

alculquicondor commented:
/release-note-edit
```release-note
Options to allow workloads to borrow quota or preempt other workloads before trying the next flavor in the list
```

alculquicondor commented on pkg/cache/snapshot_test.go:
```diff
@@ -28,19 +28,21 @@ import (
 	cmpopts.IgnoreUnexported(ClusterQueue{}),
-	cmpopts.IgnoreFields(ClusterQueue{}, "RGByResource"),
-	cmpopts.IgnoreFields(Cohort{}, "Members"), // avoid recursion.
+	cmpopts.IgnoreFields(ClusterQueue{}, "RGByResource", "Generation"),
+	cmpopts.IgnoreFields(Cohort{}, "Members", "Generation"), // avoid recursion.
```
maybe it's worth testing that the generation has the value that we expect

KunWuLuan replied:
we will only add this in snapshot cmpopts for preemption. this field will not be changed because we will not modified this field in snapshot.
we have add test to ensure this field not change in snapshot test

alculquicondor commented on pkg/features/kube_features.go:
```diff
@@ -33,6 +33,13 @@ const (
+
+	// owner: @KunWuLuan
+	// kep: https://github.com/kubernetes-sigs/kueue/tree/main/keps/582-preempt-based-on-flavor-order
+	// alpha: v0.6
```
I think we can start at beta, enabled by default

alculquicondor commented on pkg/features/kube_features.go:
```diff
@@ -33,6 +33,13 @@ const (
+	// kep: https://github.com/kubernetes-sigs/kueue/tree/main/keps/582-preempt-based-on-flavor-order
+	// alpha: v0.6
+	//
+	// Enables partial admission.
```
copy paste error :)

alculquicondor commented on pkg/features/kube_features_test.go:
```diff
@@ -25,8 +25,12 @@ import (
 
 func TestFeatureGate(t *testing.T) {
 	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, PartialAdmission, true)()
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, FlavorFungibility, true)()
```
No need to test this for every single feature

alculquicondor commented on test/integration/scheduler/scheduler_test.go:
```diff
@@ -1075,6 +1075,43 @@ var _ = ginkgo.Describe("Scheduler", func() {
 			util.ExpectWorkloadsToBeAdmitted(ctx, k8sClient, devCQ.Name, dWl1)
 		})
+
+		ginkgo.It("Should schedule when fungibility is set", func() {
```
try to be a more descriptive

alculquicondor commented on test/integration/scheduler/scheduler_test.go:
```diff
@@ -1075,6 +1075,43 @@ var _ = ginkgo.Describe("Scheduler", func() {
+			ginkgo.By("Creating another workload")
+			wl2 := testing.MakeWorkload("wl-2", ns.Name).Queue(podQueue.Name).Request(corev1.ResourceCPU, "9").Toleration(spotToleration).Obj()
+			gomega.Expect(k8sClient.Create(ctx, wl2)).Should(gomega.Succeed())
+			util.ExpectWorkloadsToBeAdmitted(ctx, k8sClient, devCQ.Name, wl2)
```
We should probably check that it's admitted in the flavor that we expect.
You can use `ExpectWorkloadToBeAdmittedAs`

KunWuLuan replied:
Done

alculquicondor commented on test/integration/scheduler/scheduler_test.go:
```diff
@@ -1075,6 +1075,43 @@ var _ = ginkgo.Describe("Scheduler", func() {
+
+			devCQ = testing.MakeClusterQueue("dev-cq").
+				Cohort("all").
+				FlavorFungibility(kueue.FlavorFungibility{WhenCanBorrow: kueue.TryNextFlavor}).
```
Add another test for the other preemption policy.

KunWuLuan replied:
sure. done

alculquicondor commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -218,33 +221,60 @@ func (m FlavorAssignmentMode) String() string {
 // FlavorAssignmentMode.
 func AssignFlavors(log logr.Logger, wl *workload.Info, resourceFlavors map[kueue.ResourceFlavorReference]*kueue.ResourceFlavor, cq *cache.ClusterQueue, counts []int32) Assignment {
+	if wl.LastAssignment != nil {
+		if cq.Generation > wl.LastAssignment.ClusterQueueGeneration &&
+			(cq.Cohort == nil || cq.Cohort.Generation > wl.LastAssignment.CohortGeneration) {
```
Shouldn't it be like this?
```suggestion
		if cq.Generation > wl.LastAssignment.ClusterQueueGeneration ||
			(cq.Cohort != nil && cq.Cohort.Generation > wl.LastAssignment.CohortGeneration) {
```

Please make sure to have a unit test for this.

KunWuLuan replied:
Sure. I will add a unit test for this.

alculquicondor commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -218,33 +221,60 @@ func (m FlavorAssignmentMode) String() string {
+		if cq.Cohort != nil {
+			assignment.LastState.CohortGeneration = cq.Cohort.Generation
+		}
 	}
```
why not keep the code as is, and add
```golang
if lastAssignment != nil {
  assigment.LastState = ...
} else {
  assignment.LastState = ...
}
```

it's less repeated code and more clear what is the actual difference.

KunWuLuan replied:
Fixed

alculquicondor reviewed (commented):
A few more comments, but I see that there are a few comments still open.

trasc commented on pkg/cache/clusterqueue.go:
```diff
@@ -281,9 +300,12 @@ func (c *ClusterQueue) updateWorkloadUsage(wi *workload.Info, m int64) {
 		updateUsage(wi, c.localQueues[qKey].usage, m)
 		c.localQueues[qKey].admittedWorkloads += int(m)
 	}
+	if m < 0 {
```
I think that the idea is change the generation only when a workload is removed, therefore well have better chances to fit in a previously checked flavor. If this is the case maybe the increment should be moved in `func (c *ClusterQueue) deleteWorkload(w *kueue.Workload) {` , and rename `Generation`  to something that describes this.

trasc commented on pkg/cache/snapshot.go:
```diff
@@ -36,8 +38,10 @@ func (s *Snapshot) RemoveWorkload(wl *workload.Info) {
 	cq := s.ClusterQueues[wl.ClusterQueue]
 	delete(cq.Workloads, workload.Key(wl.Obj))
 	updateUsage(wl, cq.Usage, -1)
+	atomic.AddInt64(&cq.Generation, 1)
```
+1 on not changing the generation in the snapshot

trasc commented on pkg/scheduler/flavorassigner/flavorassigner_test.go:
```diff
@@ -1585,16 +1805,81 @@ func TestAssignFlavors(t *testing.T) {
+				},
+			},
+			want: true,
+		},
```
add a `want=false` test

trasc commented on pkg/scheduler/flavorassigner/flavorassigner_test.go:
```diff
@@ -1585,16 +1805,81 @@ func TestAssignFlavors(t *testing.T) {
 			}
 
-			if diff := cmp.Diff(tc.wantAssignment, assignment, cmpopts.IgnoreUnexported(Assignment{}, FlavorAssignment{})); diff != "" {
+			// assignment.LastState = workload.AssigmentClusterQueueState{
+			// 	LastAssignedFlavorIdx: nil,
+			// 	ResourceFlavors:        nil,
+			// 	ClusterQueueUsage:      nil,
+			// 	CohortUsage:            nil,
+			// }
```
remove this

trasc commented on pkg/scheduler/scheduler_test.go:
```diff
@@ -1263,6 +1264,361 @@ func TestEntryOrdering(t *testing.T) {
+			}
+		})
+	}
+
```
rm

alculquicondor commented:
The integration test is failing

alculquicondor commented on test/integration/scheduler/scheduler_test.go:
```diff
@@ -1075,6 +1077,87 @@ var _ = ginkgo.Describe("Scheduler", func() {
+		ginkgo.It("Should preempt before try next flavor", func() {
+			prodCQ = testing.MakeClusterQueue("prod-cq").
+				Cohort("all").
+				ResourceGroup(*testing.MakeFlavorQuotas("on-demand").Resource(corev1.ResourceCPU, "10", "10").Obj(), *testing.MakeFlavorQuotas("spot-tainted").Resource(corev1.ResourceCPU, "11").Obj()).
```
```suggestion
				ResourceGroup(
					*testing.MakeFlavorQuotas("on-demand").Resource(corev1.ResourceCPU, "10", "10").Obj(),
					*testing.MakeFlavorQuotas("spot-tainted").Resource(corev1.ResourceCPU, "11").Obj()).
```

But maybe we can simplify and use spot-untainted? So that we don't need the toleration.

alculquicondor commented on test/integration/scheduler/scheduler_test.go:
```diff
@@ -1075,6 +1077,87 @@ var _ = ginkgo.Describe("Scheduler", func() {
+			podQueue := testing.MakeLocalQueue("dev-queue", ns.Name).ClusterQueue(devCQ.Name).Obj()
+			gomega.Expect(k8sClient.Create(ctx, podQueue)).Should(gomega.Succeed())
+
+			ginkgo.By("Creating one workload")
```
there are more than one :)

KunWuLuan replied:
Yes, thanks for reminding me.

alculquicondor commented on test/integration/scheduler/scheduler_test.go:
```diff
@@ -1075,6 +1077,87 @@ var _ = ginkgo.Describe("Scheduler", func() {
+			ginkgo.By("Creating another workload")
+			wl3 := testing.MakeWorkload("wl-3", ns.Name).Queue(prodQueue.Name).Request(corev1.ResourceCPU, "5").Toleration(spotToleration).Obj()
+			gomega.Expect(k8sClient.Create(ctx, wl3)).Should(gomega.Succeed())
+			util.ExpectWorkloadsToBePreempted(ctx, k8sClient, wl1)
```
are there any guarantees that wl1 would be preempted, instead of wl2? Maybe you can play with priorities to give some guarantees.

alculquicondor commented on test/integration/scheduler/scheduler_test.go:
```diff
@@ -1075,6 +1077,87 @@ var _ = ginkgo.Describe("Scheduler", func() {
+			devCQ = testing.MakeClusterQueue("dev-cq").
+				Cohort("all").
+				FlavorFungibility(kueue.FlavorFungibility{WhenCanBorrow: kueue.TryNextFlavor}).
+				ResourceGroup(*testing.MakeFlavorQuotas("on-demand").Resource(corev1.ResourceCPU, "10", "10").Obj(), *testing.MakeFlavorQuotas("spot-tainted").Resource(corev1.ResourceCPU, "11").Obj()).
```
```suggestion
				ResourceGroup(
					*testing.MakeFlavorQuotas("on-demand").Resource(corev1.ResourceCPU, "10", "10").Obj(),
					*testing.MakeFlavorQuotas("spot-tainted").Resource(corev1.ResourceCPU, "11").Obj()).
```

alculquicondor commented on test/integration/scheduler/scheduler_test.go:
```diff
@@ -1075,6 +1077,87 @@ var _ = ginkgo.Describe("Scheduler", func() {
+			prodQueue := testing.MakeLocalQueue("prod-queue", ns.Name).ClusterQueue(prodCQ.Name).Obj()
+			gomega.Expect(k8sClient.Create(ctx, prodQueue)).Should(gomega.Succeed())
+
+			podQueue := testing.MakeLocalQueue("dev-queue", ns.Name).ClusterQueue(devCQ.Name).Obj()
```
```suggestion
			devQueue := testing.MakeLocalQueue("dev-queue", ns.Name).ClusterQueue(devCQ.Name).Obj()
```

alculquicondor commented on test/integration/scheduler/scheduler_test.go:
```diff
@@ -1075,6 +1077,87 @@ var _ = ginkgo.Describe("Scheduler", func() {
+			prodQueue := testing.MakeLocalQueue("prod-queue", ns.Name).ClusterQueue(prodCQ.Name).Obj()
+			gomega.Expect(k8sClient.Create(ctx, prodQueue)).Should(gomega.Succeed())
+
+			podQueue := testing.MakeLocalQueue("dev-queue", ns.Name).ClusterQueue(devCQ.Name).Obj()
```
```suggestion
			devQueue := testing.MakeLocalQueue("dev-queue", ns.Name).ClusterQueue(devCQ.Name).Obj()
```

KunWuLuan commented on pkg/cache/clusterqueue.go:
```diff
@@ -23,16 +23,18 @@ var (
+	ResourceGroups                  []ResourceGroup
+	RGByResource                    map[corev1.ResourceName]*ResourceGroup
+	Usage                           workload.FlavorResourceQuantities
+	AllocatableResourceIncreasedGen int64
```
done

KunWuLuan commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -389,18 +433,62 @@ func (a *Assignment) findFlavorForResourceGroup(
 	}
+
+	for _, assignment := range bestAssignment {
+		// When the workload needborrowing and there is another workload in cohort doesn't
+		// need borrowing, the workload needborrowing will come again. In this case we should
+		// not skip the previous flavors.
+		if flavorIdx == len(rg.Flavors)-1 || assignment.Mode == Fit {
```
I set `assignment.FlavorIdx = -1` when `assignment.Mode == Fit` so that any workload need borrowing or fit comes again, it will not skip any flavors. (`assignment.FlavorIdx = -1` means no flavor would be skipped.)

set workload.LastAssigment to nil when a workload is skipped in scheduler.go should do the same thing, I think.

alculquicondor replied:
I prefer we do the changes in scheduler.go, just for separation of concerns (the flavor assigner shouldn't know about skips).

KunWuLuan replied:
Ok, I understand. Have moved it to scheduler.go

trasc commented on pkg/cache/clusterqueue.go:
```diff
@@ -51,11 +53,13 @@ type ClusterQueue struct {
 type Cohort struct {
 	Name    string
 	Members sets.Set[*ClusterQueue]
+	// This field will only be set in snapshot. This field equal to the maximum
+	// allocatable generation among its members.
+	AllocatableResourceGeneration int64
```
The problem with this approach is that if one of the non maximum gen is changing, the event will not be visible at cohort level.  Summing the might work better, Also making sure that the Gen of a queue starts from 1 will change the cohort Gen when new queues are added to it. 

Since the way the value is computed is described , it makes sense to keep this in the `These fields are only populated for a snapshot` section.

KunWuLuan replied:
Done. Because cache will call `updateResourceGroups`, Gen of a queue starts from 1

trasc commented on pkg/cache/snapshot.go:
```diff
@@ -78,13 +74,16 @@ func (c *Cache) Snapshot() Snapshot {
 				cqCopy.accumulateResources(cohortCopy)
 				cqCopy.Cohort = cohortCopy
 				cohortCopy.Members.Insert(cqCopy)
+				if cqCopy.AllocatableResourceGeneration > cohortCopy.AllocatableResourceGeneration {
+					cohortCopy.AllocatableResourceGeneration = cqCopy.AllocatableResourceGeneration
+				}
```
```suggestion
				cohortCopy.AllocatableResourceGeneration += cqCopy.AllocatableResourceGeneration
```

trasc commented on pkg/features/kube_features.go:
```diff
@@ -54,7 +54,7 @@ func init() {
 var defaultFeatureGates = map[featuregate.Feature]featuregate.FeatureSpec{
 	PartialAdmission:  {Default: false, PreRelease: featuregate.Alpha},
-	FlavorFungibility: {Default: false, PreRelease: featuregate.Alpha},
+	FlavorFungibility: {Default: true, PreRelease: featuregate.Alpha},
```
```suggestion
	FlavorFungibility: {Default: true, PreRelease: featuregate.Beta},
```

trasc commented on pkg/features/kube_features.go:
```diff
@@ -33,6 +33,13 @@ const (
+
+	// owner: @KunWuLuan
+	// kep: https://github.com/kubernetes-sigs/kueue/tree/main/keps/582-preempt-based-on-flavor-order
+	// alpha: v0.5
```
```suggestion
	// beta: v0.5
```

alculquicondor commented on pkg/cache/clusterqueue.go:
```diff
@@ -316,6 +332,7 @@ func (c *ClusterQueue) deleteWorkload(w *kueue.Workload) {
 	if c.podsReadyTracking && !apimeta.IsStatusConditionTrue(w.Status.Conditions, kueue.WorkloadPodsReady) {
 		c.WorkloadsNotReady.Delete(k)
 	}
+	c.AllocatableResourceIncreasedGen++
```
Are you saying that if the only change that happen in the snapshot was a workload add, then there is no reason to try the previous flavors again, because we know the workload won't fit. Correct?

Regardless of whether my assumption is correct, please add a comment in the code to explain the rationale.

KunWuLuan replied:
done

trasc commented on pkg/cache/clusterqueue.go:
```diff
@@ -50,8 +55,11 @@ type Cohort struct {
-	Usage                FlavorResourceQuantities
+	RequestableResources workload.FlavorResourceQuantities
+	Usage                workload.FlavorResourceQuantities
+	// This field will only be set in snapshot. This field equal to the maximum
+	// allocatable generation among its members.
```
```suggestion
	// This field will only be set in snapshot. This field equal to the sum of
	// allocatable generation among its members.
```

trasc reviewed (commented):
Otherwise LGTM

alculquicondor commented on pkg/scheduler/scheduler.go:
```diff
@@ -192,6 +192,9 @@ func (s *Scheduler) schedule(ctx context.Context) {
 			if cycleCohortsUsage.hasCommonFlavorResources(cq.Cohort.Name, e.assignment.Usage) && !cq.Cohort.CanFit(sum) {
 				e.status = skipped
 				e.inadmissibleMsg = "other workloads in the cohort were prioritized"
+				// When the workload needborrowing and there is another workload in cohort doesn't
```
```suggestion
				// When the workload needs borrowing and there is another workload in cohort doesn't
```

alculquicondor reviewed (commented):
I apologize. I forgot to send my pending comments yesterday.

KunWuLuan commented on pkg/scheduler/scheduler_test.go:
```diff
@@ -1263,6 +1264,360 @@ func TestEntryOrdering(t *testing.T) {
+			deletedWorkloads: []kueue.Workload{},
+			wantPreempted:    sets.Set[string]{},
+			wantAdmissionsOnceSchedule: map[string]kueue.Admission{
+				"default/workload": *utiltesting.MakeAdmission("eng-cohort-theta").Assignment(corev1.ResourceCPU, "spot", "20").Obj(),
```
done

alculquicondor commented on pkg/scheduler/scheduler_test.go:
```diff
@@ -1263,6 +1264,393 @@ func TestEntryOrdering(t *testing.T) {
+		workloads                      []kueue.Workload
+		deletedWorkloads               []kueue.Workload
+		wantPreempted                  sets.Set[string]
+		wantAdmissionsOnFirstSchedule  map[string]kueue.Admission
+		wantAdmissionsOnSecondSchedule map[string]kueue.Admission
```
Based on what we are comparing below
```suggestion
		wantScheduledOnFirstSchedule  map[string]kueue.Admission
		wantAdmissionsAfterSecondSchedule map[string]kueue.Admission
```

alculquicondor reviewed (commented):
/approve
/hold for 2 nits

@tenzen-y do you want to give a sanity check or are you ok with this

alculquicondor commented:
To clarify, I'm leaving the LGTM to @tenzen-y.

@KunWuLuan thanks a lot for your great effort! This was high quality work, which is even more impressive as this is your first contribution to the project. Kudos!

tenzen-y commented:
> To clarify, I'm leaving the LGTM to @tenzen-y.
> 
> @KunWuLuan thanks a lot for your great effort! This was high quality work, which is even more impressive as this is your first contribution to the project. Kudos!

Sure. Let me check this PR.

KunWuLuan commented:
> To clarify, I'm leaving the LGTM to @tenzen-y.
> 
> @KunWuLuan thanks a lot for your great effort! This was high quality work, which is even more impressive as this is your first contribution to the project. Kudos!

😆 It's really great to contribute to the Kueue community. thanks

tenzen-y commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -389,18 +433,60 @@ func (a *Assignment) findFlavorForResourceGroup(
 		}
 	}
+
+	for _, assignment := range bestAssignment {
+		if flavorIdx == len(rg.Flavors)-1 {
+			// we have reach the last flavor, try from the first flavor next time
+			assignment.FlavorIdx = -1
+		} else {
+			assignment.FlavorIdx = flavorIdx
+		}
+	}
+	if bestAssignmentMode == Fit {
+		return bestAssignment, nil
+	}
```
Are these needed if the `FlavorFungibility` featuregate is disabled?

KunWuLuan replied:
done

tenzen-y reviewed (commented):
@KunWuLuan Thank you for the great work! I left a few comments for the first pass.

Also, could you update the featuregate table?

https://github.com/kubernetes-sigs/kueue/blob/main/site/content/en/docs/installation/_index.md#change-the-feature-gates-configuration

KunWuLuan commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -218,33 +221,56 @@ func (m FlavorAssignmentMode) String() string {
+	if wl.LastAssignment != nil && lastAssignmentOutdated(wl, cq) {
+		wl.LastAssignment = nil
+	}
+
```
We will record the last flavor index in `wl.LastAssignment` and start from there next time. When `cq.AllocatebleResourceGeneration` is outdated, we set `wl.LastAssignment = nil` so that we will start from the first flavor.

tenzen-y replied:
That makes sense. Thank you for clarifying :)

KunWuLuan commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -218,33 +221,56 @@ func (m FlavorAssignmentMode) String() string {
+	} else {
+		assignment.LastState = workload.AssigmentClusterQueueState{
+			LastAssignedFlavorIdx:  make([]map[corev1.ResourceName]int, 0, len(podSets)),
+			CohortGeneration:       0,
+			ClusterQueueGeneration: cq.AllocatableResourceGeneration,
+		}
+		if cq.Cohort != nil {
+			assignment.LastState.CohortGeneration = cq.Cohort.AllocatableResourceGeneration
+		}
```
`cq.Cohort` is set in `cache.SnapShot()`, if the cluster queue is not the member of any cohort, `cq.Cohort` would be nil

tenzen-y replied:
Ah, it's correct. I was confused `cq.AllocatableResourceGeneration` with `cq.Cohort.AllocatableResourceGeneration`.

KunWuLuan commented on pkg/scheduler/flavorassigner/flavorassigner.go:
```diff
@@ -32,18 +32,21 @@ import (
 	"k8s.io/component-helpers/scheduling/corev1/nodeaffinity"
 	"k8s.io/utils/ptr"
 
+	"sigs.k8s.io/kueue/apis/kueue/v1beta1"
```
done

KunWuLuan commented on pkg/cache/snapshot_test.go:
```diff
@@ -41,6 +42,7 @@ var snapCmpOpts = []cmp.Option{
 }
 
 func TestSnapshot(t *testing.T) {
+	defer features.SetFeatureGateDuringTest(t, features.FlavorFungibility, true)()
```
this can be removed

KunWuLuan commented on pkg/cache/cache_test.go:
```diff
@@ -737,10 +797,15 @@ func TestCacheClusterQueueOperations(t *testing.T) {
+					FlavorFungibility: kueue.FlavorFungibility{
+						WhenCanBorrow:  kueue.Borrow,
+						WhenCanPreempt: kueue.TryNextFlavor,
+					},
```
done

KunWuLuan commented on pkg/cache/cache_test.go:
```diff
@@ -711,10 +766,15 @@ func TestCacheClusterQueueOperations(t *testing.T) {
+					FlavorFungibility: kueue.FlavorFungibility{
+						WhenCanBorrow:  kueue.Borrow,
+						WhenCanPreempt: kueue.TryNextFlavor,
+					},
```
done

KunWuLuan commented on pkg/cache/cache_test.go:
```diff
@@ -686,10 +736,15 @@ func TestCacheClusterQueueOperations(t *testing.T) {
+					FlavorFungibility: kueue.FlavorFungibility{
+						WhenCanBorrow:  kueue.Borrow,
+						WhenCanPreempt: kueue.TryNextFlavor,
+					},
```
done

KunWuLuan commented on pkg/cache/clusterqueue.go:
```diff
@@ -167,6 +174,18 @@ func (c *ClusterQueue) update(in *kueue.ClusterQueue, resourceFlavors map[kueue.
+			c.FlavorFungibility.WhenCanBorrow = kueue.Borrow
+		}
+		if c.FlavorFungibility.WhenCanPreempt == "" {
+			c.FlavorFungibility.WhenCanPreempt = kueue.TryNextFlavor
```
done

KunWuLuan commented on pkg/cache/clusterqueue.go:
```diff
@@ -167,6 +174,18 @@ func (c *ClusterQueue) update(in *kueue.ClusterQueue, resourceFlavors map[kueue.
+	if in.Spec.FlavorFungibility != nil {
+		c.FlavorFungibility = *in.Spec.FlavorFungibility
+		if c.FlavorFungibility.WhenCanBorrow == "" {
+			c.FlavorFungibility.WhenCanBorrow = kueue.Borrow
```
done, thanks

KunWuLuan commented on apis/kueue/v1beta1/clusterqueue_types.go:
```diff
@@ -274,6 +278,40 @@ const (
+	//   candidates for preemption in the current flavor.
+	//
+	// +kubebuilder:validation:Enum={Preempt,TryNextFlavor}
+	// +kubebuilder:default="TryNextFlavor"
```
When `whenCanPreempt=Preempt`, it means that we will pause at each flavor where usage+request<nominal+borrowingLimit and attempt to preempt. When `whenCanPreempt=tryNextFlavor`, it means that we will try to preempt after traversing all flavors and when mode=Preempt. `whenCanPreempt=tryNextFlavor` is the current default behavior because we will traverse all flavors and break the loop until the best mode is `Fit` or all flavors have been .

tenzen-y replied:
Ah, right. The flavorassigner traverses all flavors if the mode is `Preempt`.
Thank you for the clarifications, and Sorry for the confusion.

https://github.com/kubernetes-sigs/kueue/blob/c4197436d17e6241cd71d880c68d4329fa2f65c8/pkg/scheduler/flavorassigner/flavorassigner.go#L392-L399

tenzen-y commented on pkg/cache/cache_test.go:
I'm still waiting for this.

KunWuLuan replied:
done

tenzen-y commented on pkg/util/testing/wrappers.go:
```diff
@@ -318,6 +318,14 @@ func MakeClusterQueue(name string) *ClusterQueueWrapper {
 		Spec: kueue.ClusterQueueSpec{
 			NamespaceSelector: &metav1.LabelSelector{},
 			QueueingStrategy:  kueue.BestEffortFIFO,
+			Preemption: &kueue.ClusterQueuePreemption{
```
I'm still waiting for this.

KunWuLuan replied:
done. thanks

tenzen-y commented on pkg/webhooks/clusterqueue_webhook.go:
```diff
@@ -67,6 +67,18 @@ func (w *ClusterQueueWebhook) Default(ctx context.Context, obj runtime.Object) e
+			WhenCanPreempt: kueue.TryNextFlavor,
+		}
+	}
+	if cq.Spec.FlavorFungibility.WhenCanBorrow == "" {
+		cq.Spec.FlavorFungibility.WhenCanBorrow = kueue.Borrow
+	}
+	if cq.Spec.FlavorFungibility.WhenCanPreempt == "" {
+		cq.Spec.FlavorFungibility.WhenCanPreempt = kueue.TryNextFlavor
+	}
```
I don't think we need these defaulters since the crd defaulters inject these default values into these, right?

tenzen-y reviewed (commented):
I reviewed everything.

tenzen-y commented:
I'm still waiting for some comments.

tenzen-y commented on pkg/cache/cache_test.go:
```diff
@@ -739,11 +793,13 @@ func TestCacheClusterQueueOperations(t *testing.T) {
+					Preemption:                    defaultPreemption,
+					AllocatableResourceGeneration: 1,
+					FlavorFungibility:             defaultFlavorFungibility,
+					AdmissionChecks:               sets.Set[string]{"check1": {}, "check2": {}},
```
```suggestion
					AdmissionChecks:               sets.New("check1", "check2"),
```

KunWuLuan replied:
done

tenzen-y commented:
> hi @KunWuLuan we are approaching the time for the next release. Can you follow up on the last suggestions?

I'm still waiting for the following:

- https://github.com/kubernetes-sigs/kueue/pull/849#discussion_r1333997290
- https://github.com/kubernetes-sigs/kueue/pull/849#discussion_r1334055881
- https://github.com/kubernetes-sigs/kueue/pull/849#discussion_r1334145099
- https://github.com/kubernetes-sigs/kueue/pull/849#discussion_r1341047332

KunWuLuan commented:
> > hi @KunWuLuan we are approaching the time for the next release. Can you follow up on the last suggestions?
> 
> I'm still waiting for the following:
> 
> * [Enable preemption and borrowing based on flavor #849 (comment)](https://github.com/kubernetes-sigs/kueue/pull/849#discussion_r1333997290)
> * [Enable preemption and borrowing based on flavor #849 (comment)](https://github.com/kubernetes-sigs/kueue/pull/849#discussion_r1334055881)
> * [Enable preemption and borrowing based on flavor #849 (comment)](https://github.com/kubernetes-sigs/kueue/pull/849#discussion_r1334145099)
> * [Enable preemption and borrowing based on flavor #849 (comment)](https://github.com/kubernetes-sigs/kueue/pull/849#discussion_r1341047332)

Sure, for replying late. I am on vacation this week. I will work on these comments this weekend.

tenzen-y commented:
> > > hi @KunWuLuan we are approaching the time for the next release. Can you follow up on the last suggestions?
> > 
> > 
> > I'm still waiting for the following:
> > 
> > * [Enable preemption and borrowing based on flavor #849 (comment)](https://github.com/kubernetes-sigs/kueue/pull/849#discussion_r1333997290)
> > * [Enable preemption and borrowing based on flavor #849 (comment)](https://github.com/kubernetes-sigs/kueue/pull/849#discussion_r1334055881)
> > * [Enable preemption and borrowing based on flavor #849 (comment)](https://github.com/kubernetes-sigs/kueue/pull/849#discussion_r1334145099)
> > * [Enable preemption and borrowing based on flavor #849 (comment)](https://github.com/kubernetes-sigs/kueue/pull/849#discussion_r1341047332)
> 
> Sure, for replying late. I am on vacation this week. I will work on these comments this weekend.

Thank you!

KunWuLuan commented:
> > > > hi @KunWuLuan we are approaching the time for the next release. Can you follow up on the last suggestions?
> > > 
> > > 
> > > I'm still waiting for the following:
> > > 
> > > * [Enable preemption and borrowing based on flavor #849 (comment)](https://github.com/kubernetes-sigs/kueue/pull/849#discussion_r1333997290)
> > > * [Enable preemption and borrowing based on flavor #849 (comment)](https://github.com/kubernetes-sigs/kueue/pull/849#discussion_r1334055881)
> > > * [Enable preemption and borrowing based on flavor #849 (comment)](https://github.com/kubernetes-sigs/kueue/pull/849#discussion_r1334145099)
> > > * [Enable preemption and borrowing based on flavor #849 (comment)](https://github.com/kubernetes-sigs/kueue/pull/849#discussion_r1341047332)
> > 
> > 
> > Sure, for replying late. I am on vacation this week. I will work on these comments this weekend.
> 
> Thank you!

Hi, I have addressed these comments. Please take a look if u have time. Thanks very much.

tenzen-y reviewed (commented):
@KunWuLuan Thank you for this great work!

/lgtm
/approve

tenzen-y commented:
/hold cancel
